runtime/near-vm-logic/src/logic.rs

```diff
@@ -943,14 +945,117 @@ impl<'a> VMLogic<'a> {
-        self.gas_counter.pay_per_byte(keccak512_byte, value.len() as u64)?;
+        self.gas_counter.pay_per(keccak512_byte, value.len() as u64)?;
 
         use sha3::Digest;
 
         let value_hash = sha3::Keccak512::digest(&value);
         self.internal_write_register(register_id, value_hash.as_slice().to_vec())
     }
 
+    /// Hashes the given value using RIPEMD-160 and returns it into `register_id`.
+    ///
+    /// # Errors
+    ///
+    /// If `value_len + value_ptr` points outside the memory or the registers use more memory than
+    /// the limit with `MemoryAccessViolation`.
+    ///
+    /// # Cost
+    ///
+    ///  Where `message_blocks` is `(value_len + 9).div_ceil(64)`.
+    ///
+    /// `base + write_register_base + write_register_byte * num_bytes + ripemd160_base + ripemd160_block * message_blocks`
+    #[cfg(feature = "protocol_feature_math_extension")]
+    pub fn ripemd160(&mut self, value_len: u64, value_ptr: u64, register_id: u64) -> Result<()> {
+        self.gas_counter.pay_base(ripemd160_base)?;
+        let value = self.get_vec_from_memory_or_register(value_ptr, value_len)?;
+
+        let message_blocks = value
+            .len()
+            .checked_add(8)
+            .ok_or(VMLogicError::HostError(HostError::IntegerOverflow))?
+            / 64
+            + 1;
+
+        self.gas_counter.pay_per(ripemd160_block, message_blocks as u64)?;
+
+        use ripemd160::Digest;
+
+        let value_hash = ripemd160::Ripemd160::digest(&value);
+        self.internal_write_register(register_id, value_hash.as_slice().to_vec())
+    }
+
+    /// Recovers an ECDSA signer address and returns it into `register_id`.
+    ///
+    /// Takes in an additional flag to check for malleability of the signature
+    /// which is generally only ideal for transactions.
+    ///
+    /// Returns a bool indicating success or failure as a `u64`.
+    ///
+    /// # Malleability Flags
+    ///
+    /// 0 - No malleability check.
+    /// 1 - Check malleability.
+    /// 2 - Rejecting upper range.
+    ///
+    /// # Errors
+    ///
+    /// * If `hash_ptr`, `r_ptr`, or `s_ptr` point outside the memory or the registers use more
+    ///   memory than the limit, then returns `MemoryAccessViolation`.
+    ///
+    /// # Cost
+    ///
+    /// `base + write_register_base + write_register_byte * 20 + ecrecover_base`
+    #[cfg(feature = "protocol_feature_math_extension")]
+    pub fn ecrecover(
+        &mut self,
+        hash_ptr: u64,
+        sig_ptr: u64,
+        v: u64,
+        malleability_flag: u64,
+        register_id: u64,
+    ) -> Result<u64> {
+        self.gas_counter.pay_base(ecrecover_base)?;
+
+        let signature = {
+            let vec = self.get_vec_from_memory_or_register(sig_ptr, 64)?;
+            let mut bytes = [0u8; 65];
+            bytes[0..64].copy_from_slice(&vec);
+
+            match v {
+                0 | 1 | 2 | 3 => {
+                    bytes[65] = v as u8;
```

Comment:
```suggestion
                    bytes[64] = v as u8;
```

65 is out of bounds I think?

Reply:
65 bytes, so 0 -> 64, then yes. Just got back from picking up my son from his cousins, getting this in now.